rtos/source/Semaphore.cpp

```diff
@@ -104,43 +108,53 @@ void Semaphore::acquire()
 }
 
 bool Semaphore::try_acquire_for(uint32_t millisec)
+{
+    return try_acquire_for(duration<uint32_t, milli>(millisec));
+}
+
+bool Semaphore::try_acquire_for(Kernel::Clock::duration_u32 rel_time)
 {
 #if MBED_CONF_RTOS_PRESENT
-    osStatus_t status = osSemaphoreAcquire(_id, millisec);
+    osStatus_t status = osSemaphoreAcquire(_id, rel_time.count());
     if (status == osOK) {
         return true;
     }
 
-    bool success = (status == osErrorResource && millisec == 0) ||
-                   (status == osErrorTimeout && millisec != osWaitForever);
+    bool success = (status == osErrorResource && rel_time == rel_time.zero()) ||
+                   (status == osErrorTimeout);
```

Comment:
I believe I was just trying to simplify the "pickiness". Yes, that is the only time "timeout" should occur, but I felt we don't need to be "asserting" that.

"Resource" errors can occur for two different reasons - one is "not available with zero timeout" and the other is "this semaphore handle is invalid". So it's worth trying to catch and make the second scenario fault, rather than swallowing it.

 Does that make sense?

Reply:
Yes, I agree, we can simplify the  "pickiness". 
`Mutex::trylock_for` could have the same change then - my comment here was mostly triggered by the difference between the two files.